cmake/Version.cmake

```
@@ -8,20 +8,100 @@
 
 string(TIMESTAMP current_year "%Y")
 set(O3DE_COPYRIGHT_YEAR ${current_year} CACHE STRING "Open 3D Engine's copyright year")
-set(O3DE_VERSION_STRING "0.0.0" CACHE STRING "Open 3D Engine's version")
-set(O3DE_DISPLAY_VERSION_STRING "00.00" CACHE STRING "Open 3D Engine's display version")
-set(O3DE_BUILD_VERSION 0 CACHE STRING "Open 3D Engine's build number")
-set(O3DE_ENGINE_NAME "o3de" CACHE STRING "Open 3D Engine's engine name")
-
-# Optional environment overrides
-if(NOT "$ENV{O3DE_VERSION}" STREQUAL "")
-    set(O3DE_VERSION_STRING "$ENV{O3DE_VERSION}")
-endif()
-
-if(NOT "$ENV{O3DE_DISPLAY_VERSION}" STREQUAL "")
-    set(O3DE_DISPLAY_VERSION_STRING "$ENV{O3DE_DISPLAY_VERSION}")
-endif()
-
-if(NOT "$ENV{O3DE_BUILD_VERSION}" STREQUAL "")
-    set(O3DE_BUILD_VERSION "$ENV{O3DE_BUILD_VERSION}")
-endif()
+
+# avoid reading engine.json multiple times 
+ly_file_read("${LY_ROOT_FOLDER}/engine.json" tmp_json_data)
+set_property(GLOBAL PROPERTY O3DE_ENGINE_JSON_DATA ${tmp_json_data})
+unset(tmp_json_data)
+
+#! o3de_read_engine_default: Read a field from engine.json or use the default if not found
+# \arg:output_value - name of output variable to set 
+# \arg:key - name of field in engine.json 
+# \arg:default_value - value to use if neither environment var or engine.json value found 
+macro(o3de_read_engine_default output_value key default_value)
+    get_property(engine_json_data GLOBAL PROPERTY O3DE_ENGINE_JSON_DATA)
+    string(JSON tmp_value ERROR_VARIABLE manifest_json_error GET ${engine_json_data} ${key})
+
+    # unset engine_json_data because we're in a macro
+    unset(${engine_json_data})
+
+    if(manifest_json_error)
+        message(WARNING "Failed to read ${key} from \"${LY_ROOT_FOLDER}/engine.json\" : ${manifest_json_error}")
+        set(${tmp_value} ${default_value})
```

repeat: This should reference tmp_value which is the name of the variable, not `${tmp_value}` which is the value of the variable.
```suggestion
        set(tmp_value ${default_value})
```